src/System.Private.CoreLib/src/System/Threading/ClrThreadPool.WaitThread.cs

```diff
@@ -0,0 +1,395 @@
+            }
+
+            /// <summary>
+            /// Starts the wait thread if it has not been started yet.
+            /// </summary>
+            private void StartWaitThreadIfNotStarted()
+            {
+                if(_waitThreadStarted)
+                {
+                    return;
+                }
+
+                _waitThreadStartedLock.Acquire();
+                if (!_waitThreadStarted)
+                {
+                    _waitHandles[0] = _changeHandlesEvent;
+                    RuntimeThread waitThread = RuntimeThread.Create(WaitThreadStart);
+                    waitThread.IsBackground = true;
+                    waitThread.Start();
+                    _waitThreadStarted = true;
+                }
+                _waitThreadStartedLock.Release();
+            }
+
+            /// <summary>
+            /// Queues on the thread pool or executes directly a call to <see cref="UnregisterWait(object)"/>.
+            /// </summary>
+            /// <param name="handle">The handle to unregister.</param>
+            /// <remarks>
+            /// As per CoreCLR's behavior, if the user passes in a <see cref="WaitHandle"/> that is equal to <c>-1</c>
+            /// into <see cref="RegisteredWaitHandle.Unregister(WaitHandle)"/>, then the unregistration of the wait handle is blocking.
+            /// Otherwise, the unregistration of the wait handle is queued on the thread pool.
+            /// </remarks>
+            public void QueueOrExecuteUnregisterWait(RegisteredWaitHandle handle)
+            {
+                if (handle.UserUnregisterWaitHandle?.SafeWaitHandle.DangerousGetHandle() == (IntPtr)(-1))
+                {
+                    UnregisterWait(handle);
+                }
+                else
+                {
+                    ThreadPool.QueueUserWorkItem(UnregisterWait, handle);
+                }
+                // Wait until it is safe for the user to dispose the wait handle they just unregistered before returning.
+                _safeToDisposeHandleEvent.WaitOne();
+            }
+
+            /// <summary>
+            /// Unregister a wait handle.
+            /// </summary>
+            /// <param name="state">The wait handle to unregister.</param>
+            private void UnregisterWait(object state)
+            {
+                RegisteredWaitHandle handle = (RegisteredWaitHandle)state;
+                
+                // TODO: Optimization: Try to unregister wait directly if it isn't being waited on.
+                _removesLock.Acquire();
+                if(Array.IndexOf(_pendingRemoves, handle) == -1) // If this handle is not already pending removal
+                {
+                    _pendingRemoves[_numPendingRemoves++] = handle;
+                }
+                _removesLock.Release();
+
+                _changeHandlesEvent.Set(); // Tell the wait thread that there are changes pending.
+                if (handle.UserUnregisterWaitHandle != null)
```

Comment:
`RegisteredWaitHandle` internally ensures that the user-provided wait handle is only signaled once and the callback is not called if the handle has been signaled.